In [1]:
from sklearn.datasets import load_iris
import numpy as np

iris = load_iris()
X = iris.data      # matriz de características
y = iris.target    # etiquetas: 0, 1, 2
print(X.shape)
print(iris.feature_names)

(150, 4)
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [2]:
x0 = X[0]          # primera muestra: un vector
print(x0)
print(x0.shape)

[5.1 3.5 1.4 0.2]
(4,)


In [3]:
# fila = muestra ; columna = característica
col_2 = X[:, 2]    # 3ª característica de TODAS las muestras
print(col_2[:5])
print(col_2.shape)

[1.4 1.4 1.3 1.5 1.4]
(150,)


In [4]:
print(X.shape)     # (150, 4)
print(X.T.shape)   # transpuesta: (4, 150)

(150, 4)
(4, 150)


In [5]:
a = X[0]           # una flor setosa
b = X[50]          # una flor versicolor
print('a + b =', a + b)
print('2 * a =', 2 * a)

a + b = [12.1  6.7  6.1  1.6]
2 * a = [10.2  7.   2.8  0.4]


In [6]:
manual = sum(ai * bi for ai, bi in zip(a, b))
print('a mano :', round(manual, 2))
print('np.dot :', round(np.dot(a, b), 2))
print('con @  :', round(a @ b, 2))

a mano : 53.76
np.dot : 53.76
con @  : 53.76


In [7]:
# ADVERTENCIA: a * b NO es producto punto, es multiplicación elemento a elemento
print('a * b (elemento a elemento) =', a * b)

a * b (elemento a elemento) = [35.7  11.2   6.58  0.28]


In [8]:
def coseno(u, v):
    return np.dot(u, v) / (np.linalg.norm(u) * np.linalg.norm(v))

print('cos(a, b) =', round(coseno(a, b), 4))

cos(a, b) = 0.9284


In [9]:
setosa1    = X[0]
setosa2    = X[1]
versicolor = X[50]
print('cos(setosa1, setosa2)    =', round(coseno(setosa1, setosa2), 4))
print('cos(setosa1, versicolor) =', round(coseno(setosa1, versicolor), 4))

cos(setosa1, setosa2)    = 0.9986
cos(setosa1, versicolor) = 0.9284


In [10]:
def distancia(u, v):
    return np.linalg.norm(u - v)

print('d(setosa1, setosa2)    =', round(distancia(setosa1, setosa2), 4))
print('d(setosa1, versicolor) =', round(distancia(setosa1, versicolor), 4))

d(setosa1, setosa2)    = 0.5385
d(setosa1, versicolor) = 4.0037


In [11]:
x = X[0]
print('x . x      =', round(np.dot(x, x), 4))
print('||x|| ** 2 =', round(np.linalg.norm(x) ** 2, 4))

x . x      = 40.26
||x|| ** 2 = 40.26


## 07 · Entregable evaluable

Resuelto con NumPy (solo `np.dot` y `np.linalg.norm`); sin funciones de distancia/similitud de scikit-learn.

In [12]:
# 1) Cargar iris y seleccionar 3 muestras:
#    i = 0  -> setosa     (y[0]  = 0)
#    j = 1  -> setosa     (y[1]  = 0)  (misma especie que i)
#    k = 50 -> versicolor (y[50] = 1)  (especie distinta)
idx = [0, 1, 50]
muestras = X[idx]
print('Indices     :', idx)
print('Etiquetas   :', y[idx])
print('Matriz 3x4 de muestras:\n', muestras)

Indices     : [0, 1, 50]
Etiquetas   : [0 0 1]
Matriz 3x4 de muestras:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [7.  3.2 4.7 1.4]]


In [13]:
# 2) Matriz 3x3 de distancias euclidianas
D = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        D[i, j] = distancia(muestras[i], muestras[j])
print('Matriz de distancias euclidianas 3x3:\n', D)

Matriz de distancias euclidianas 3x3:
 [[0.         0.53851648 4.00374824]
 [0.53851648 0.         4.09633983]
 [4.00374824 4.09633983 0.        ]]


In [14]:
# 3) Matriz 3x3 de similitud coseno
C = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        C[i, j] = coseno(muestras[i], muestras[j])
print('Matriz de similitud coseno 3x3:\n', C)

Matriz de similitud coseno 3x3:
 [[1.         0.99857916 0.92838036]
 [0.99857916 1.         0.94000276]
 [0.92838036 0.94000276 1.        ]]


In [15]:
# 4) Verificar para muestras[0] que ||x||^2 = x . x
xv = muestras[0]
print('x . x      =', round(np.dot(xv, xv), 4))
print('||x|| ** 2 =', round(np.linalg.norm(xv) ** 2, 4))
print('diferencia =', round(np.dot(xv, xv) - np.linalg.norm(xv) ** 2, 12))

x . x      = 40.26
||x|| ** 2 = 40.26
diferencia = 0.0


**Interpretación (3–4 líneas).**  
El par más parecido según la similitud coseno es `(setosa1, setosa2)` con coseno ≈ 0.9984, y también es el más cercano según la distancia euclidiana (0.5385). Ambas métricas coinciden porque ambas muestras son de la misma especie, por lo que sus vectores apuntan en direcciones casi idénticas y están cerca en el espacio. La distancia euclidiana depende también de la magnitud del vector, mientras que el coseno solo de la dirección; por eso pueden diferir cuando cambia el tamaño pero se mantiene la forma.